### PR TITLE
Implementation of Istio mixer adapter

### PR DESCRIPTION
**Pull request by HasithaAthukorala**

### CODE REVIEW

File: components/global/core/io.cellery.observability.telemetry.receiver/src/main/java/io/cellery/observability/telemetry/receiver/internal/MetricsHandler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package io.cellery.observability.telemetry.receiver.internal;
```
Comment (afkham) [inline components/global/core/io.cellery.observability.telemetry.receiver/src/main/java/io/cellery/observability/telemetry/receiver/internal/MetricsHandler.java]: No license header. This indicates that your have't configured your IDE to automatically insert the header

File: components/global/core/io.cellery.observability.telemetry.receiver/pom.xml [hunk 2]
```diff
@@ -34,6 +34,12 @@
             <groupId>org.wso2.siddhi</groupId>
             <artifactId>siddhi-core</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.sun.net.httpserver</groupId>
+            <artifactId>http</artifactId>
+            <version>20070405</version>
```
Comment (nadundesilva) [inline components/global/core/io.cellery.observability.telemetry.receiver/pom.xml]: Please inherit from the root pom and use the existing dependency.

File: components/global/core/io.cellery.observability.telemetry.receiver/pom.xml [hunk 3]
```diff
@@ -96,6 +102,12 @@
             <artifactId>commons-io</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (nadundesilva) [inline components/global/core/io.cellery.observability.telemetry.receiver/pom.xml]: Let's use Gson instead of this since it is already added to SP.

File: components/global/core/io.cellery.observability.siddhi.apps/src/main/siddhi/istio-telemetry-app.siddhi [hunk 4]
```diff
@@ -112,7 +113,8 @@ from K8sComponentPodsInMemorySink
 select name as podName, cell, component
 insert into K8sComponentPodInfoWindow;
 
-from TelemetryStream[(not sourceUID is null) and (not destinationUID is null)]
+from TelemetryStream
+[(not sourceUID is null) and (not destinationUID is null)]
```
Comment (nadundesilva) [inline components/global/core/io.cellery.observability.siddhi.apps/src/main/siddhi/istio-telemetry-app.siddhi]: Let's revert this change

File: components/global/mixer-adapter/cmd/mixer-adapter/main.go [hunk 5]
```diff
@@ -14,35 +14,76 @@
...
+	defer func() {
+		err := logger.Sync()
+		if err != nil {
+			log.Fatalf("Error syncing logger: %s", err.Error())
+		}
+	}()
+
+	/* Mutual TLS feature to secure connection between workloads
+	   This is optional. */
+	credential := os.Getenv(grpcAdapterCredentialPath)   // adapter.crt //change the name
```
Comment (nadundesilva) [inline components/global/mixer-adapter/cmd/mixer-adapter/main.go]: Let's rename `credential` to `certificate` to be more meaningful.

Also rename `certificate` variable below to `caCertificate`

Please rename in all the relevant places.

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go [hunk 6]
```diff
@@ -60,53 +52,94 @@ type (
...
+
+	/* This interface and the struct has been implemented to test the publishMetrics() function */
+	MetricsPublisher interface {
+		publishMetrics(attributeMap map[string]interface{}, logger *zap.SugaredLogger, httpClient *http.Client, spServerUrl string) bool
 	}
+
+	SPMetricsPublisher struct{}
 )
 
 // HandleMetric records metric entries
```
Comment (nadundesilva) [inline components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go]: This comment is misleading. Let's update it.

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go [hunk 7]
```diff
@@ -60,53 +52,94 @@ type (
...
 
-		defer file.Close()
+	return &v1beta1.ReportResult{}, nil
+}
 
-		if _, err = file.Write(buffer.Bytes()); err != nil {
-			wso2SpAdapter.logger.Error("error writing to file: ", err.Error())
-			return nil, err
-		}
+/* Send metrics to sp server */
```
Comment (nadundesilva) [inline components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go]: Let's stick the standard go style doc comments.

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go [hunk 8]
```diff
@@ -60,53 +52,94 @@ type (
...
+		return false
+	}
+	defer func() {
+		err := resp.Body.Close()
+		if err != nil {
+			logger.Warn("Could not close the body")
+		}
+	}()
+	body, _ := ioutil.ReadAll(resp.Body)
+	logger.Info("response : ", string(body))
```
Comment (nadundesilva) [inline components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go]: This should be a debug log and it should be something more meaningful.

Suggestion :- "Received response from SP Worker: " + string(body)

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go [hunk 9]
```diff
@@ -143,80 +182,46 @@ func instances(in []*metric.InstanceMsg) string {
...
-		ClientCAs:    certPool,
-	}
-	tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
-
-	return grpc.Creds(credentials.NewTLS(tlsConfig)), nil
-}
-
-// NewWso2SpAdapter creates a new IBP adapter that listens at provided port.
-func NewWso2SpAdapter(addr string, logger *zap.SugaredLogger) (Server, error) {
+// New creates a new IBP adapter that listens at provided port.
```
Comment (nadundesilva) [inline components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go]: Shouldn't this be `creates a new SP adapter`

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter_test.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+	var sampleInstances []*metric.InstanceMsg
+	sampleInstances = append(sampleInstances, sampleInstance1)
+	sampleInstances = append(sampleInstances, sampleInstance2)
+	sampleInstances = append(sampleInstances, sampleInstance3)
+	sampleInstances = append(sampleInstances, sampleInstance4)
+
+	anyType := &types.Any{}
+	configP := &config.Params{}
+	configP.FilePath = "out.txt"
```
Comment (nadundesilva) [inline components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter_test.go]: We don't need any config file path now right ?

File: docker/mixer-adapter/Dockerfile [hunk 11]
```diff
@@ -20,4 +20,6 @@ FROM scratch
 
 COPY ./target/mixer-adapter /mixer-adapter
 
+ENV SP_SERVER_URL "http://wso2sp-worker.cellery-system.svc.cluster.local:9091"
```
Comment (nadundesilva) [inline docker/mixer-adapter/Dockerfile]: We don't need to add this here I think. This is something we can directly est from the deployment yaml. The default value we are setting is anyway a Kubernetes specific value.

File: components/global/mixer-adapter/cmd/mixer-adapter/main.go [hunk 12]
```diff
@@ -14,35 +14,76 @@
...
+	privateKey := os.Getenv(grpcAdapterPrivateKeyPath)   // adapter.key
+	certificate := os.Getenv(grpcAdapterCertificatePath) // ca.pem
+	spServerUrl := os.Getenv(spServerUrl)
+
+	logger.Infof("Sp server url : %s", spServerUrl)
+
+	client := &http.Client{}
+	publisher := wso2spadapter.SPMetricsPublisher{}
```
Comment (Mirage20) [inline components/global/mixer-adapter/cmd/mixer-adapter/main.go]: Do we need SP in the struct? can't we use MetricsPublisher{}? Its anyway in the wso2spadapter package.

File: components/global/mixer-adapter/cmd/mixer-adapter/main.go [hunk 13]
```diff
@@ -14,35 +14,76 @@
...
+		if err != nil {
+			logger.Warn("Server option could not be fetched, Connection will not be encrypted")
+		}
 	}
-	defer logger.Sync()
 
-	adapter, err := wso2spadapter.NewWso2SpAdapter(addr, logger)
+	adapter, err := wso2spadapter.New(port, logger, client, publisher, serverOption, spServerUrl)
 	if err != nil {
 		logger.Error("unable to start server: ", err.Error())
```
Comment (Mirage20) [inline components/global/mixer-adapter/cmd/mixer-adapter/main.go]: you can use logger.Fatal instead of using logger.Error and os.Exit

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go [hunk 14]
```diff
@@ -60,53 +52,94 @@ type (
...
+		server           *grpc.Server
+		logger           *zap.SugaredLogger
+		httpClient       *http.Client
+		metricsPublisher MetricsPublisher
+		spServerUrl      string
+	}
+
+	/* This interface and the struct has been implemented to test the publishMetrics() function */
+	MetricsPublisher interface {
+		publishMetrics(attributeMap map[string]interface{}, logger *zap.SugaredLogger, httpClient *http.Client, spServerUrl string) bool
```
Comment (Mirage20) [inline components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go]: Either make publishMetrics public or make MetricsPublisher private. Otherwise, we are exposing an empty interface

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go [hunk 15]
```diff
@@ -60,53 +52,94 @@ type (
...
 	}
+
+	SPMetricsPublisher struct{}
```
Comment (Mirage20) [inline components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go]: Rename to MetricsPublisher

File: components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go (hunk 14 above)
Comment (Mirage20) [reply components/global/mixer-adapter/pkg/wso2spadapter/wso2spadapter.go]: You can create interface similar to this
Publisher interface {
   Publish(...) bool
}

File: components/global/mixer-adapter/cmd/mixer-adapter/main.go [hunk 16]
```diff
@@ -19,38 +19,112 @@
...
+		if err != nil {
+			log.Fatalf("Error syncing logger: %s", err.Error())
+		}
+	}()
 
 	if len(os.Args) > 1 {
-		addr = os.Args[1]
+		port, err = strconv.Atoi(os.Args[1])
+		if err != nil {
+			logger.Errorf("Could not convert the port number from string to int : ", err.Error())
```
Comment (azinneera) [inline components/global/mixer-adapter/cmd/mixer-adapter/main.go]: `%s` missing in the error message

File: components/global/core/io.cellery.observability.telemetry.receiver/src/main/java/io/cellery/observability/telemetry/receiver/internal/MetricsHandler.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    private static final Logger log = Logger.getLogger(MetricsHandler.class);
+    private static final Gson gson = new Gson();
+    private SourceEventListener sourceEventListener;
+
+    public MetricsHandler(SourceEventListener sourceEventListener) {
+        this.sourceEventListener = sourceEventListener;
+    }
+
+    @Override
+    public void handle(HttpExchange httpExchange) throws IOException {
```
Comment (azinneera) [inline components/global/core/io.cellery.observability.telemetry.receiver/src/main/java/io/cellery/observability/telemetry/receiver/internal/MetricsHandler.java]: You can use try-with-resources for the OutputStream and InputStreamReader

File: components/global/mixer-adapter/cmd/mixer-adapter/main.go (hunk 16 above)
Comment (azinneera) [inline components/global/mixer-adapter/cmd/mixer-adapter/main.go]: `%s` missing in the error message

File: components/global/mixer-adapter/pkg/adapter/adapter.go [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+		adapter.publisher.Publish(attributesMap, adapter.logger, adapter.httpClient, adapter.spServerUrl) // ToDO : get the boolean value from the function to check whether the metric is delivered or not
+	}
+
+	return &v1beta1.ReportResult{}, nil
+}
+
+// Send metrics to sp server
+func (publisher SPMetricsPublisher) Publish(attributeMap map[string]interface{}, logger *zap.SugaredLogger, httpClient *http.Client, spServerUrl string) bool {
+	jsonValue, _ := json.Marshal(attributeMap)
+	req, err := http.NewRequest("POST", spServerUrl, bytes.NewBuffer(jsonValue))
```
Comment (azinneera) [inline components/global/mixer-adapter/pkg/adapter/adapter.go]: We should handle the error here.
